src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -3053,13 +3173,29 @@ makeLimit (conn, _) (Limit ml mo) =
         v = maybe 0 fromIntegral
     in (TLB.fromText limitRaw, mempty)
 
-makeLocking :: LockingClause -> (TLB.Builder, [PersistValue])
-makeLocking = flip (,) [] . maybe mempty toTLB . Monoid.getLast
+makeLocking :: IdentInfo -> LockingClause -> (TLB.Builder, [PersistValue])
+makeLocking _ (GeneralLockingClause lockingClause) =
+  case lockingClause of
+    ForUpdate           -> ("\nFOR UPDATE", [])
+    ForUpdateSkipLocked -> ("\nFOR UPDATE SKIP LOCKED", [])
+    ForShare            -> ("\nFOR SHARE", [])
+    LockInShareMode     -> ("\nLOCK IN SHARE MODE", [])
+makeLocking info (PostgresLockingClause (PostgresLockingKind rowStrength onLockBehaviour)) = 
+  makeLockingStrength info rowStrength <> plain " " <> makeLockingBehavior onLockBehaviour
   where
-    toTLB ForUpdate           = "\nFOR UPDATE"
-    toTLB ForUpdateSkipLocked = "\nFOR UPDATE SKIP LOCKED"
-    toTLB ForShare            = "\nFOR SHARE"
-    toTLB LockInShareMode     = "\nLOCK IN SHARE MODE"
+    makeLockingStrength :: IdentInfo -> PostgresRowLevelLockStrength -> (TLB.Builder, [PersistValue])
+    makeLockingStrength info (PostgresForUpdate Nothing) = plain "FOR UPDATE" 
+    makeLockingStrength info (PostgresForUpdate (Just (Of lockableEnts))) =
+      plain "FOR UPDATE OF " <> makeLockableEntity info lockableEnts
+    makeLockingStrength info (PostgresForShare Nothing) = plain "FOR SHARE"
+    makeLockingStrength info (PostgresForShare (Just (Of lockableEnts))) =
+      plain "FOR SHARE OF " <> makeLockableEntity info lockableEnts
+    makeLockingBehavior :: OnLockedBehavior -> (TLB.Builder, [PersistValue])
+    makeLockingBehavior NoWait = plain "NO WAIT"
+    makeLockingBehavior SkipLocked = plain "SKIP LOCKED"
+    makeLockingBehavior Wait = plain ""
+    plain v = (v,[])
+makeLocking _ NoLockingClause = mempty
```

Comment:
So, if we're gonna make this 'extensible', you'd probably need this to be a class member -

```haskell
class ToLockingClause a where
    renderLockingClause :: IdentInfo -> a -> (TLB.Builder, [PersistValue])

data SomeLockingClause where
    SomeLockingClause :: ToLockingClause a => a -> SomeLockingClause
```

Then the stuff in `Internal` can be "generic SQL", and `.Postgres` can define

```haskell
data PostgresqlLockingClause

instance ToLockingClause PostgresqlLockingClause
```

But that's not quite enough. We need a way of determining the "strength", or a way of comparing two locking clauses. Otherwise we can't `<>` two things together.

And comparing two things of potentially disparate types is not really an easy problem to solve!

🤔

Reply:
So I tried implementing this but it makes the monoid instance impossible to define I think since I can't pattern match on it :(

I've made it a class member but this function still needs to pattern match to destructure  it.. Not sure if there's a good way of using the existential type and having a monoid instance